### DISCUSSION TITLE
Unexpected behaviour when using AWS SDK v3 with Vercel

### DISCUSSION THREAD

**Discussion by choutkamartin** [vercel/community#88] [Help] [answered]
Hi, 

I started this discussion because I think that this is not necessarily an in issue type, nor does it relate to Next.js too much.

The problem is that on [this page](https://vercel.com/support/articles/serverless-functions-and-smtp) is written we should use AWS SES as a mail service.

So, I used SES in my project. I've come across that it's a must to use different naming for environment variables to avoid conflict with Vercel ones. It's written in [this article](https://vercel.com/support/articles/how-can-i-use-aws-sdk-environment-variables-on-vercel). 

If we do the above and use AWS SDK v2, everything will work on localhost and on Vercel. 

However, Amazon also introduced AWS SDK v3, which is modularized, for example, [this is the SES package](https://www.npmjs.com/package/@aws-sdk/client-ses). If we use the v3 package and do just a few things differently, we will come to a conclusion, that v3 will of course send email just fine - on localhost. It sadly won't work on Vercel. The v3 SDK throws out this error: 
```
Error occured: AccessDenied: User `arn:aws:sts::867210282167:assumed-role/cloudwatch_logs_events_putter/5HW69rHjvWBIaHTxG9cOTBQI-23f50e60d4541359d4fe4483f966b678b68f355' is not authorized to perform `ses:SendEmail' on resource `arn:aws:ses:eu-central-1:867210282167:identity/************'
```

In the log above is no AWS ID that is mine. Both user and resource have ID 867210282167, maybe it's the ID of the Vercel lambdas?

You can checkout repository, and the live running example here: https://github.com/choutkamartin/next-ses

The account is in a sandbox, and the emails are sent to the AWS simulator email address, so you can send a few test emails until the sandbox hits a limit. But don't overuse it though, thanks.

**Comment by leerob** [comment]
It's possible the v3 version changed the IAM permission needed to use SES - the "Access Denied" message isn't an error from Vercel, but from AWS.

**Comment by choutkamartin** [comment, marked as answer]
@leerob Hi Lee, thanks for the answer. It was not due to the permissions. It was due to AWS picking up Vercel's environment variables instead of the mine defined, as I inputted them wrong into the client. Why it worked on localhost is because the client works weirdly. I will analyze the behaviour and won't discuss it until I get confirmation from the SDK team that it is indeed a feature and not a bug.

Because Next.js doesn't have any example using the v2 or v3 version, I will create one under the Next.js repository. Also, the v3 version doesn't have a simple example in the AWS docs. 🙂

**Reply by leerob** [reply]
Glad it's resolved! Thank you for all the info.

**Reply by kyduff** [reply]
@choutkamartin can you elaborate on how you resolved the issue (or provide a link to your solution)? I've encountered an identical problem with my Next deployment (using AWS v3 SDKs) today.

**Reply by choutkamartin** [reply]
@kyduff  I will provide a detailed solution in an hour, gimme a minute 🙂

**Comment by choutkamartin** [comment]
@kyduff The problem on my side was the way I passed the credentials to the client like this:
```js
const credentials = {
  accessKeyId: process.env.AWS_SDK_ACCESS_KEY_ID,
  secretAccessKey: process.env.AWS_SDK_SECRET_ACCESS_KEY,
  region: process.env.AWS_SDK_REGION,
};

const s3Client = new S3Client(credentials);
```
But the correct way is:
```js
const credentials = {
  accessKeyId: process.env.AWS_SDK_ACCESS_KEY_ID,
  secretAccessKey: process.env.AWS_SDK_SECRET_ACCESS_KEY,
};

const s3Client = new S3Client({
  region: process.env.AWS_SDK_REGION,
  credentials: credentials,
});

// Or if you use SES
const sesClient = new SESClient({
  region: process.env.AWS_SDK_REGION,
  credentials: credentials,
});
```

In the wrong example, the client was picking up the environment variables Vercel uses (as Vercel deploys projects on AWS). https://vercel.com/docs/concepts/limits/overview#reserved-variables

In my case, sending emails worked fine on *localhost* (even though the API key was passed to the client wrongly). That's because, even though it seems strange, an API key is not needed when you are sending mail from an email address to the same one (and the email address is verified).

Also, make sure that you created an IAM user with correct permission policies (in the case of this example AmazonS3FullAccess, and AmazonSESFullAccess) and generated the access key ID and secret access key for the user.

**Reply by kyduff** [reply]
@choutkamartin Thanks so much for this write-up! I'm passing the environment variables in the first (wrong) way, so I'll make the change and hope it results in a fix. Will keep you updated.

Update: this worked!

**Reply by choutkamartin** [reply]
@kyduff You're welcome 🙂

**Reply by arturmuller** [reply]
Had similar issue, your code example brought me on the right track -- thanks so much!
